### ISSUE TITLE
Rather confusing help command

### ISSUE THREAD

**Issue by flip111** [symfony/symfony#20030]
This is from the symfony installer:

```
» symfony --help
Usage:
  help [options] [--] [<command_name>]

Arguments:
  command               The command to execute
  command_name          The command name [default: "help"]

Options:
      --xml             To output help as XML
      --format=FORMAT   The output format (txt, xml, json, or md) [default: "txt"]
      --raw             To output raw command help
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi            Force ANSI output
      --no-ansi         Disable ANSI output
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
 The help command displays help for a given command:

   php /home/flip111/.local/bin/symfony help list

 You can also output the help in other formats by using the --format option:

   php /home/flip111/.local/bin/symfony help --format=xml list

 To display the list of available commands, please use the list command.
```
1. `[--]` looks pretty cryptic, doesn't really indicate what it does
2. `Arguments` is not on the first line `help [options] [--] [<command_name>]`. If command_name is part of arguments then options is also an argument.
3. `command` is not part of the first line
4. `command_name` why do i need to give it a name when i already supply the name at `command`?
5. `Help` .. I'm already in the `help` command, What's `Help` in `help`.
6. `command` is missing from the help, it should really be there without running again `list`
7. `format` option is explained twice

Then running `list`

```
» symfony list
Symfony Installer version 1.5.8

Usage:
  command [options] [arguments]

Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi            Force ANSI output
      --no-ansi         Disable ANSI output
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Available commands:
  about        Symfony Installer Help.
  demo         Creates a demo Symfony project.
  help         Displays help for a command
  list         Lists commands
  new          Creates a new Symfony project.
  self-update  Update the Symfony Installer to the latest version.
  selfupdate   Update the Symfony Installer to the latest version.
```
1. i get a usage again, but the usage of what? All commands? How do the options and arguments distinct with the usage of a command and the usage of the executable?
2. Options of what? Since the executable has some of the same options should i type like `php bin/console --ansi command --ansi` ? Or are these options for the executable and thus for all commands rather than a specific command?

What's also rather random is that in the second paste i get the name + version number. Which is omitted in the first paste ..

Overall i find the experience not as concise as some other command-line utilities help. Perhaps merge `help` and `list` into `help`.

**Comment by HeahDude** [comment]
Can you please open this issue in [symfony/symfony-installer](https://github.com/symfony/symfony-installer/issues) and close this one? Thanks!

**Comment by wouterj** [comment]
@HeahDude as you can see, this help is generated by Symfony Console. `symfony --help` is only used as an example to get the help output.

**Comment by ro0NL** [comment]
> [--] looks pretty cryptic, doesn't really indicate what it does

AFAIK this is a common CLI signature (separates options from arguments). Could be noted somewhere..

> Arguments is not on the first line help [options] [--] [<command_name>]. If command_name is part of arguments then options is also an argument.

No.. after `[--]` comes arguments :) `[options]` is a shortcut for `--someoption --etc --long-stufff`.

> command is not part of the first line

True.. it's a global application argument (opposed to command arguments) which shouldnt be here, fixed in #20054 and #19181

> command_name why do i need to give it a name when i already supply the name at command?

This is an argument of the `help` command, ie you get help via `help my_command_name`. This could get a better description though (opposed to `The command name`).

> Help .. I'm already in the help command, What's Help in help.

`help` is a command.. you can get help for it like any other command.
```
help == help help == help -h == help --help
help list == list -h == list --help
```

> command is missing from the help, it should really be there without running again list

Not sure i follow.. if you're talking about `"command               The command to execute"`; it shouldnt be here.

> format option is explained twice

I see it only once `--format=FORMAT   The output format `. However it can look confusing opposed to `--raw` and `--xml`.

> i get a usage again, but the usage of what? All commands? How do the options and arguments distinct with the usage of a command and the usage of the executable?

You get a usage on how to run a command... not sure this really makes sense (you expect the usage of the `list` command).

Options of what? Since the executable has some of the same options should i type like php bin/console --ansi command --ansi ? Or are these options for the executable and thus for all commands rather than a specific command?

All commands :)

**Comment by flip111** [comment]
> No.. after [--] comes arguments :)

That's why it's confusing because after `[--]` comes `[<command_name>]`.

> This is an argument of the help command

When i do `--help` i expect help on the entire console app, not help on help.

> help is a command.. you can get help for it like any other command

I would expect to get this with the following syntax: `symfony --help help`, see also previous point that using `--help` i expect information on the app.

> if you're talking about "command The command to execute"; it shouldnt be here

`--help` should give information on which commands there are and a short description of them, otherwise it's pretty useless help (that mainly seems to give help on itself, see previous 2 points).

> I see it only once --format=FORMAT The output format

I see also `You can also output the help in other formats by using the --format option`

> You get a usage on how to run a command... not sure this really makes sense

No it doesn't because that information should come when you do `console --help` not `console list`

**Comment by wouterj** [comment]
> That's why it's confusing because after [--] comes [<command_name>].

It's just a normal CLI convention that unix and everybody else use. Not worth changing it imo. See also http://docopt.org/

**Comment by flip111** [comment]
> It's just a normal CLI convention that unix and everybody else use. Not worth changing it imo. See also http://docopt.org/

I'm just saying `"arguments" !== "command_name"` don't know what's CLI convention about that ...

**Comment by xabbuh** [comment]
@flip111 Well, in your example `command_name` is an argument of the help command.

**Comment by wouterj** [comment]
So, let's make a few things clear about the current situation:

 #. `symfony new --help` is effectively the same as `symfony help new`
 #. `symfony --help` is equal to `symfony help --help`.

So what we're seeing here is the help info of the `help` command, not of the complete `symfony` application.

To me, the behaviour of (1) seems perfectly fine and it's what I expect to be happening.

However, I think @flip111 has done a nice job in explaning that (2) is *not* the expected behaviour. I would expect `symfony --help` to show the help of the application.

I propose to do the following:

 * If `Application#getHelp()` equals `Application#getLongVersion()` (i.e. the `getHelp()` method is not overridden), `symfony --help` should show the output returned by `Application#getHelp()`.
 * Otherwise, execute the default command (which is probably `list`) and show this output in `symfony --help`. This means `symfony` and `symfony --help` behave the same, which seems kinda logical to me.

**Comment by flip111** [comment]
@wouterj that would be a good thing to make it more intuitive.

> `symfony --help` is equal to `symfony help --help`.

I didn't realise this actually since you mention it.

**Comment by javiereguiluz** [comment]
Closing this discussion because it stalled and also, we haven't received any other complaints about this from the community (and that's important because Console is the second most downloaded Symfony component with 150 million downloads, so thousands of people use it in all kinds of applications). Thanks.

**Comment by ro0NL** [comment]
i think the only wrong issue here is

```
Arguments:
  command               The command to execute
  command_name          The command name [default: "help"]
```

which is reported in #17804 also
